File: VB/GrowthWare.Framework.BusinessData/DataAccessLayer/MySQL/V5_6_21/DRoles.vb
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.MySql.Base
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
Imports MySql.Data.MySqlClient
Imports GrowthWare.Framework.Model.Profiles

Namespace DataAccessLayer.MySql.V5621
    Public Class DRoles
        Inherits DDBInteraction
        Implements IDRoles

        Public Sub DeleteRole() Implements IDRoles.DeleteRole
            Dim mStoredProcedure As String = "ZGWSecurity.Delete_Role"
            Dim mParameters() As MySqlParameter =
            {
             New MySqlParameter("@P_Name", Profile.Name),
             New MySqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqId)
            }
            MyBase.ExecuteNonQuery(mStoredProcedure, mParameters)
        End Sub

        Public Function GetAccountsInRole() As DataTable Implements IDRoles.AccountsInRole
            Dim mStoreProcedure As String = "ZGWSecurity.Get_Accounts_In_Role"
            Dim mParameters() As MySqlParameter =
            {
              New MySqlParameter("@P_Role_SeqID", Profile.Id),
              New MySqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqId)
            }
            Return MyBase.GetDataTable(mStoreProcedure, mParameters)
        End Function

        Public Function GetAccountsNotInRole() As DataTable Implements IDRoles.AccountsNotInRole
            Dim mStoreProcedure As String = "ZGWSecurity.Get_Accounts_Not_In_Role"
            Dim mParameters() As MySqlParameter =
            {
             New MySqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqId),
             New MySqlParameter("@P_Role_SeqID", Profile.Id)
            }
            Return MyBase.GetDataTable(mStoreProcedure, mParameters)
        End Function

        Public Function GetProfileData() As DataRow Implements IDRoles.ProfileData
            Dim mStoredProcedure As String = "ZGWSecurity.Get_Role"
            Dim mParameters() As MySqlParameter =
            {
              New MySqlParameter("@P_Role_SeqID", Profile.Id),
              New MySqlParameter("@P_Security_Entity_SeqID", -1)
            }
            Return MyBase.GetDataRow(mStoredProcedure, mParameters)
        End Function

        Public Function GetRolesBySecurityEntity() As DataTable Implements IDRoles.RolesBySecurityEntity
            Dim mStoredProcedure As String = "ZGWSecurity.Get_Role"
            Dim mParameters() As MySqlParameter =
            {
              New MySqlParameter("@P_Role_SeqID", -1),
              New MySqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqId)
            }
            Return MyBase.GetDataTable(mStoredProcedure, mParameters)
        End Function

        Public Property Profile As MRoleProfile Implements IDRoles.Profile

        Public Function Save() As Integer Implements IDRoles.Save
            Dim mRetVal As Integer
            Dim mStoredProcedure As String = "ZGWSecurity.Set_Role"
            Dim mParameters() As MySqlParameter = getInsertUpdateParameters()
            MyBase.ExecuteNonQuery(mStoredProcedure, mParameters)
            mRetVal = Integer.Parse(MyBase.GetParameterValue("@P_Primary_Key", mParameters))
            Return mRetVal
        End Function

        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable Implements IDRoles.Search
            If searchCriteria Is Nothing Then Throw New ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!")
            Dim mStoredProcedure As String = "ZGWSystem.Get_Paginated_Data"
            Dim mRetVal As DataTable = Nothing
            Dim mParameters() As MySqlParameter =
             {
              New MySqlParameter("@P_Columns", searchCriteria.Columns),
              New MySqlParameter("@P_OrderByColumn", searchCriteria.OrderByColumn),
              New MySqlParameter("@P_OrderByDirection", searchCriteria.OrderByDirection),
              New MySqlParameter("@P_PageSize", searchCriteria.PageSize),
              New MySqlParameter("@P_SelectedPage", searchCriteria.SelectedPage),
              New MySqlParameter("@P_TableOrView", "ZGWSecurity.vwSearchRoles"),
              New MySqlParameter("@P_WhereClause", searchCriteria.WhereClause)
             }
            mRetVal = MyBase.GetDataTable(mStoredProcedure, mParameters)
            Return mRetVal
        End Function

        Public Property SecurityEntitySeqId As Integer Implements IDRoles.SecurityEntitySeqId

        Public Function UpdateAllAccountsForRole(roleSeqId As Integer, securityEntityId As Integer, accounts() As String, accountSeqId As Integer) As Boolean Implements IDRoles.UpdateAllAccountsForRole
            If accounts Is Nothing Then Throw New ArgumentNullException("accounts", "accounts cannot be a null reference (Nothing in Visual Basic)!")
            Dim success As Boolean = False
            Dim db As MySqlConnection = Nothing
            Dim dbConn As MySqlCommand = Nothing
            Dim trans As MySqlTransaction = Nothing
            Dim account As String
            Try
                db = New MySqlConnection(ConnectionString)
                trans = db.BeginTransaction(IsolationLevel.Serializable)
                ' delete all the accounts for this role/SecurityEntity
                dbConn = New MySqlCommand("ZGWSecurity.Delete_Roles_Accounts", db)
                Dim myParameter As MySqlParameter = New MySqlParameter("@P_ROLE_SEQ_ID", roleSeqId)
                dbConn.Parameters.Add(myParameter)
                myParameter = New MySqlParameter("@P_Security_Entity_SeqID", securityEntityId)
                dbConn.Parameters.Add(myParameter)
                dbConn.ExecuteNonQuery()

                For Each account In accounts
                    dbConn.Parameters.Clear()
                    dbConn.CommandText = "ZGWSecurity.Set_Role_Accounts"
                    myParameter = New MySqlParameter("@P_Role_SeqID", roleSeqId)
                    dbConn.Parameters.Add(myParameter)
                    myParameter = New MySqlParameter("@P_Security_Entity_SeqID", securityEntityId)
                    dbConn.Parameters.Add(myParameter)
                    myParameter = New MySqlParameter("@P_Account", account)
                    dbConn.Parameters.Add(myParameter)
                    myParameter = New MySqlParameter("@P_Added_Updated_By", accountSeqId)
                    dbConn.Parameters.Add(myParameter)
                    dbConn.ExecuteNonQuery()
                Next
                success = True
                trans.Commit()
            Catch ex As Exception
                If Not trans Is Nothing Then trans.Rollback()
                Throw
            Finally
                If Not trans Is Nothing Then
                    trans.Dispose()
                    trans = Nothing
                End If
                If Not (db Is Nothing) Then
                    db.Close()
                End If
                If Not dbConn Is Nothing Then
                    dbConn.Dispose()
                    dbConn = Nothing
                End If
            End Try
            Return success

        End Function

        Private Function getInsertUpdateParameters() As MySqlParameter()
            Dim myParameters() As MySqlParameter =
             { _
             New MySqlParameter("@P_Role_SeqID", Profile.Id),
             New MySqlParameter("@P_Name", Profile.Name),
             New MySqlParameter("@P_Description", Profile.Description),
             New MySqlParameter("@P_Is_System", Profile.IsSystem),
             New MySqlParameter("@P_Is_System_Only", Profile.IsSystemOnly),
             New MySqlParameter("@P_Security_Entity_SeqID", SecurityEntitySeqId),
             New MySqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(Profile)),
             GetSqlParameter("@P_Primary_Key", Profile.Id, ParameterDirection.Output)
             }
            Return myParameters
        End Function

    End Class
End Namespace

File: Net_Core/DataAccess/SQLServer/DRoles.cs
using GrowthWare.DataAccess.Interfaces;
using GrowthWare.DataAccess.SQLServer.Base;
using GrowthWare.Framework.Models;
using System;
using System.Data;
using System.Data.SqlClient;
using System.Threading.Tasks;

namespace GrowthWare.DataAccess.SQLServer;

/// <summary>
/// Class DRoles.
/// </summary>
public class DRoles : AbstractDBInteraction, IRoles
{

    #region Member Fields
    private int m_SecurityEntityID;

    private MRole m_Profile = new();
    #endregion

    #region Constructors
    public DRoles(string connectionString, int securityEntitySeqId) : base()
    {
        this.ConnectionString = connectionString;
        this.m_SecurityEntityID = securityEntitySeqId;
    }
    #endregion

    int IRoles.SecurityEntitySeqID
    {
        get { return m_SecurityEntityID; }
        set { m_SecurityEntityID = value; }
    }

    MRole IRoles.Profile
    {
        get { return m_Profile; }
        set { m_Profile = value; }
    }

    async Task IRoles.DeleteRole(string roleName, int securityEntitySeqId)
    {
        SqlParameter[] mParameters = [new("@P_Name", roleName), new("@P_SecurityEntitySeqId", securityEntitySeqId)];
        String mStoreProc = "[ZGWSecurity].[Delete_Role]";
        await base.ExecuteNonQueryAsync(mStoreProc, mParameters);
    }

    async Task<int> IRoles.Save(MRole profile)
    {
        SqlParameter[] mParameters = getInsertUpdateParameters(profile);
        string myStoreProcedure = "[ZGWSecurity].[Set_Role]";
        await base.ExecuteNonQueryAsync(myStoreProcedure, mParameters);
        int mRetVal = int.Parse(GetParameterValue("@P_Primary_Key", mParameters));
        return mRetVal;
    }

    async Task<DataTable> IRoles.RolesBySecurityEntity(int securityEntitySeqId)
    {
        // a roleSeqId of -1 will return all rows for the given securityEntitySeqId
        SqlParameter[] myParameters = [new("@P_RoleSeqId", -1), new("@P_SecurityEntitySeqId", securityEntitySeqId)];
        String myStoreProc = "[ZGWSecurity].[Get_Role]";
        return await base.GetDataTableAsync(myStoreProc, myParameters);
    }

    async Task<DataRow> IRoles.ProfileData(int roleSeqId)
    {
        // a roleSeqId <> -1 will return a single row for the given roleSeqId and the securityEntitySeqId is ignored
        SqlParameter[] myParameters = [new("@P_RoleSeqId", roleSeqId), new("@P_SecurityEntitySeqId", -1)];
        String myStoreProc = "[ZGWSecurity].[Get_Role]";
        return await base.GetDataRowAsync(myStoreProc, myParameters);
    }

    async Task<DataTable> IRoles.AccountsInRole()
    {
        SqlParameter[] myParameters = [new("@P_SecurityEntitySeqId", m_SecurityEntityID), new("@P_RoleSeqId", m_Profile.Id)];
        string myStoreProcedure = "[ZGWSecurity].[Get_Accounts_In_Role]";
        return await base.GetDataTableAsync(myStoreProcedure, myParameters);
    }

    async Task<DataTable> IRoles.AccountsNotInRole()
    {
        SqlParameter[] myParameters = [new("@P_SecurityEntitySeqId", m_SecurityEntityID), new("@P_RoleSeqId", m_Profile.Id)];
        string myStoreProcedure = "[ZGWSecurity].[Get_Accounts_Not_In_Role]";
        return await base.GetDataTableAsync(myStoreProcedure, myParameters);
    }

    async Task<bool> IRoles.UpdateAllAccountsForRole(int roleSeqID, int SecurityEntityID, string[] accounts, int accountSeqID)
    {
        bool mRetVal = false;
        SqlConnection mSqlConnection = null;
        SqlTransaction mSqlTransaction = null;
        string mAccount = null;
        try
        {
            mSqlConnection = new SqlConnection(ConnectionString);
            mSqlConnection.Open();
            mSqlTransaction = mSqlConnection.BeginTransaction(IsolationLevel.Serializable);
            // delete all the accounts for this role/SecurityEntity
            SqlCommand mSqlCommand = new SqlCommand("[ZGWSecurity].[Delete_Roles_Accounts]", mSqlConnection);
            mSqlCommand.CommandType = CommandType.StoredProcedure;
            mSqlCommand.Transaction = mSqlTransaction;

            SqlParameter mSqlParameter = new("@P_ROLE_SEQ_ID", roleSeqID);
            mSqlCommand.Parameters.Add(mSqlParameter);
            mSqlParameter = new("@P_SecurityEntitySeqId", SecurityEntityID);
            mSqlCommand.Parameters.Add(mSqlParameter);
            await mSqlCommand.ExecuteNonQueryAsync();

            mSqlCommand.CommandText = "[ZGWSecurity].[Set_Role_Accounts]";
            foreach (string account_loopVariable in accounts)
            {
                mAccount = account_loopVariable;
                mSqlCommand.Parameters.Clear();
                mSqlParameter = new("@P_RoleSeqId", roleSeqID);
                mSqlCommand.Parameters.Add(mSqlParameter);
                mSqlParameter = new("@P_SecurityEntitySeqId", SecurityEntityID);
                mSqlCommand.Parameters.Add(mSqlParameter);
                mSqlParameter = new("@P_Account", mAccount);
                mSqlCommand.Parameters.Add(mSqlParameter);
                mSqlParameter = new("@P_Added_Updated_By", accountSeqID);
                mSqlCommand.Parameters.Add(mSqlParameter);
                await mSqlCommand.ExecuteNonQueryAsync();
            }

            mSqlTransaction.Commit();
            mRetVal = true;
        }
        catch (Exception)
        {
            if ((mSqlTransaction != null)) mSqlTransaction.Rollback();
            throw;
        }
        finally
        {
            if ((mSqlTransaction != null))
            {
                mSqlTransaction.Dispose();
                mSqlTransaction = null;
            }
            if ((mSqlConnection != null))
            {
                mSqlConnection.Dispose();
                mSqlConnection = null;
            }
        }
        return mRetVal;
    }

    private SqlParameter[] getInsertUpdateParameters(MRole profile)
    {
        SqlParameter[] mParameters =
        {
            new("@P_RoleSeqId", profile.Id),
            new("@P_Name", profile.Name),
            new("@P_Description", profile.Description),
            new("@P_Is_System", profile.IsSystem),
            new("@P_Is_System_Only", profile.IsSystemOnly),
            new("@P_SecurityEntitySeqId", profile.SecurityEntityID),
            new("@P_Added_Updated_By", GetAddedUpdatedBy(profile, profile.Id)),
            GetSqlParameter("@P_Primary_Key", profile.Id, ParameterDirection.Output)
        };
        return mParameters;
    }

}
